crates/biome_service/src/workspace/server.rs

```diff
@@ -181,88 +182,64 @@ impl WorkspaceServer {
         workspace_mut.set_current_project(project_key);
     }
 
-    /// Checks whether, if the current path belongs to the current project.
+    /// Checks whether the current path belongs to another project.
     ///
-    /// If there's a match, and the match **isn't** the current project, it returns the new key.
-    fn path_belongs_to_current_workspace(&self, path: &BiomePath) -> Option<ProjectKey> {
+    /// If there's a match, and the match is for a project **other than** the current project, it
+    /// returns the new key.
+    fn path_belongs_to_other_project(&self, path: &BiomePath) -> Option<ProjectKey> {
         let workspace = self.workspace();
-        workspace.as_ref().path_belongs_to_current_workspace(path)
+        workspace.as_ref().path_belongs_to_other_project(path)
     }
 
-    /// Get the parser result for a given file
+    /// Retrieves the parser result for a given file, calculating it if the file was not yet parsed.
     ///
-    /// Returns and error if no file exists in the workspace with this path or
-    /// if the language associated with the file has no parser capability
-    fn get_parse(&self, biome_path: BiomePath) -> Result<AnyParse, WorkspaceError> {
-        match self.syntax.entry(biome_path) {
-            Entry::Occupied(entry) => Ok(entry.get().clone()),
-            Entry::Vacant(entry) => {
-                let biome_path = entry.key();
-                let capabilities = self.get_file_capabilities(biome_path);
-
-                let mut document = self
-                    .documents
-                    .get_mut(biome_path)
-                    .ok_or_else(WorkspaceError::not_found)?;
-
-                let parse = capabilities
-                    .parser
-                    .parse
-                    .ok_or_else(self.build_capability_error(biome_path))?;
-
-                let document = &mut *document;
-                let workspace = self.workspace();
-                let Some(file_source) = self.get_source(document.file_source_index) else {
-                    return Err(WorkspaceError::not_found());
-                };
-                let settings = workspace.settings();
-                let parsed = parse(
-                    biome_path,
-                    file_source,
-                    document.content.as_str(),
-                    settings,
-                    &mut document.node_cache,
-                );
-                let ParseResult {
-                    language,
-                    any_parse,
-                } = parsed;
-                if let Some(language) = language {
-                    document.file_source_index = self.set_source(language);
-                }
-                Ok(entry.insert(any_parse).clone())
-            }
-        }
+    /// Returns an error if no file exists in the workspace with this path.
+    fn get_parse(&self, biome_path: &BiomePath) -> Result<AnyParse, WorkspaceError> {
+        self.documents
+            .pin()
```

Comment:
Yes, it is the equivalent of calling `.read()` or `.write()` on an `RwLock`. The pin is like a guard, but it’s called differently because it doesn’t lock the map. What it _does_ do is that it gives you mutable access to the map and it prevents any references you take from getting cleaned up as long as you keep it pinned.